src/expr/src/scalar/func/impls/jsonb.rs

```diff
@@ -33,6 +34,9 @@ sqlfunc!(
 
 sqlfunc!(
     #[sqlname = "jsonb_to_smallint"]
+    #[preserves_uniqueness = false]
```

Comment:
I think the defaults in `macro_rules! sqlfunc` are
```rust
#[preserves_uniqueness = false]
#[inverse = None]
````
so I wonder why you are explicitly adding them everywhere.

Reply:
The issue is that (AFAICT) the macro needs a separate pattern for every possible combination of annotations. Naively, this means that adding a new annotation doubles the number of necessary patterns.

I figured it was better to provide some redundant annotations than to add additional patterns to the macro, but I don't feel strongly about it; happy to go the other way if y'all prefer it.

Reply:
The `sqlfunc!` calls itself recursively top-down. With the current definition:

https://github.com/MaterializeInc/materialize/blob/ee747b414cdb3f726de72a7a4fa8bc1feb6881ba/src/expr/src/scalar/func/macros.rs#L17-L139

* the first application will just add the default `#[sqlname = stringify!($fn_name)]`,
* the second will add `#[preserves_uniqueness = false]` and `#[inverse = None]`,
* and so forth ...

If you use `rust-analyzer` you should have an "expand macro recursively" action and check what happens if you don't annotate your function explicitly with the defaults. I'm pretty sure that with your current changes you should be able to validate if this works.

Reply:
I see where the problem is, I'll try to see if there is a more elegant solution that covers `N` independent sets of defaults where any `M < N` of those can be present in some arbitrary order.

Reply:
Asked in our internal #rust-help channel... we'll see if anyone has ideas!

Reply:
I think you might be able to pattern match all attributes with `$(#[$name = $val])` but I have not tried it out.